### PR TITLE
Increase Network Policy tests coverage

### PR DESCRIPTION
**Pull request by MaysaMacedo**
**What this PR does / why we need it**:
Validates correct behavior when :

- NetworkPolicyIngressRule is updated
- Network Policy is enforced and Namespace label is updated
- Network Policy is enforced and Pod label is updated
- Network Policy is enforced based on NamespaceSelector and named port

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
/kind feature
/area testing
/sig-testing

### CODE REVIEW

**Comment by MaysaMacedo** [comment]
@kubernetes/sig-network-pr-reviews

**Comment by vllry** [comment]
/assign @vllry

Will review when I get to the office. :)

**Comment by MaysaMacedo** [comment]
Thank you, @vllry . :)

File: test/e2e/network/network_policy.go [hunk 1]
```diff
@@ -17,9 +17,11 @@ limitations under the License.
 package network
 
 import (
+	"encoding/json"
 	v1 "k8s.io/api/core/v1"
 	networkingv1 "k8s.io/api/networking/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	types "k8s.io/apimachinery/pkg/types"
```
Comment (vllry) [inline test/e2e/network/network_policy.go]: Compile error due to this import: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/78177/pull-kubernetes-bazel-build/1134114894237405184/build-log.txt

Requires explicit dependency listing, not transitive dependencies.

File: test/e2e/network/network_policy.go [hunk 2]
```diff
@@ -312,6 +314,52 @@ var _ = SIGDescribe("NetworkPolicy", func() {
 			})
 		})
 
+		ginkgo.It("should allow ingress access from namespace-b on one named port [Feature:NetworkPolicy]", func() {
```
Comment (vllry) [inline test/e2e/network/network_policy.go]: The test name should be more generic - EG `should allow ingress access from namespace on one named port`.

File: test/e2e/network/network_policy.go [hunk 3]
```diff
@@ -312,6 +314,52 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+			}
+
+			policy, err = f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Create(policy)
+			framework.ExpectNoError(err, "Error creating Network Policy %v: %v", policy.ObjectMeta.Name, err)
+			defer cleanupNetworkPolicy(f, policy)
+
+			ginkgo.By("Creating client-a which should not be able to contact the server.", func() {
+				testCannotConnect(f, f.Namespace, "client-a", service, 80)
+			})
+			ginkgo.By("Creating client-b which should not be able to contact the server on port 80.", func() {
```
Comment (vllry) [inline test/e2e/network/network_policy.go]: Incorrect test message?

File: test/e2e/network/network_policy.go [hunk 4]
```diff
@@ -353,6 +401,170 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+					}},
+				},
+			}
+			policy, err = f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Update(policy)
+			framework.ExpectNoError(err, "Error updating Network Policy %v: %v", policy.ObjectMeta.Name, err)
+			defer cleanupNetworkPolicy(f, policy)
+
+			testCanConnect(f, f.Namespace, "client-b", service, 81)
+			testCannotConnect(f, f.Namespace, "client-c", service, 81)
+			testCannotConnect(f, f.Namespace, "client-b", service, 80)
```
Comment (vllry) [inline test/e2e/network/network_policy.go]: Why use 3 clients when you only test with 2 at each stage?

File: test/e2e/network/network_policy.go [hunk 5]
```diff
@@ -353,6 +401,170 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+					Ingress: []networkingv1.NetworkPolicyIngressRule{{
+						From: []networkingv1.NetworkPolicyPeer{{
+							PodSelector: &metav1.LabelSelector{
+								MatchLabels: map[string]string{
+									"pod-name": "client-a",
+								},
+							},
+						}},
+						Ports: []networkingv1.NetworkPolicyPort{{
+							Port: &intstr.IntOrString{IntVal: 80},
```
Comment (vllry) [inline test/e2e/network/network_policy.go]: This is nitpicky, but I would prefer variables over magic numbers (EG clientAAllowedPort = 80). Makes the code easier to read without cross-referencing. Not sure if this follows or conflicts with official style though?

File: test/e2e/network/network_policy.go [hunk 6]
```diff
@@ -353,6 +401,170 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+					}},
+				},
+			}
+			policy, err := f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Create(policy)
+			framework.ExpectNoError(err, "Error creating Network Policy %v: %v", policy.ObjectMeta.Name, err)
+			testCanConnect(f, f.Namespace, "client-a", service, 80)
+			testCannotConnect(f, f.Namespace, "client-c", service, 80)
+			testCannotConnect(f, f.Namespace, "client-a", service, 81)
+
+			ginkgo.By("Updating a network policy for the Service which allows traffic from client-b at port 81.")
```
Comment (vllry) [inline test/e2e/network/network_policy.go]: Same comment about test name, the name should be the functionality tested, without any implementation details ("client-b").

**Comment by vllry** [comment]
Thanks for submitting @MaysaMacedo! There's a few small issues and the build needs fixing.

File: test/e2e/network/network_policy.go (hunk 4 above)
Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: When I used only two options of client names some errors occurred due to a client pod with same name as the one in creation  be in terminating status. But, I managed to reorganize to only use   two clients names.

Comment (vllry) [reply test/e2e/network/network_policy.go]: Is there still a concern around this timing condition? Flaking tests are a big issue in the Kubernetes codebase, so we try to either explicitly wait for events, or don't share resources.

Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: I've updated the implementation to explicit wait for the pod deletion event to be done.

File: test/e2e/network/network_policy.go (hunk 5 above)
Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: I'm not sure if there is any official style for this case, I couldn't find any. If there is some I would really appreciate if you could point me to it.
I also think is better to have variables, I was trying to follow the current style for Network Policies tests.
I took a look on some others network tests, I could find some that do not use ports as magic numbers, such as [kubeproxy](https://github.com/MaysaMacedo/kubernetes/blob/add-more-e2e-np-tests/test/e2e/network/kube_proxy.go#L44) and  [networking](https://github.com/MaysaMacedo/kubernetes/blob/add-more-e2e-np-tests/test/e2e/network/networking.go#L109).

**Comment by MaysaMacedo** [comment]
Thank you for the review @vllry .

File: test/e2e/network/network_policy.go (hunk 5 above)
Comment (vllry) [reply test/e2e/network/network_policy.go]: If you don't mind! I won't block the PR based on this, it's just a recommendation.

Unfortunately, outside of the API, there aren't very strong style guidelines that I know of.

**Comment by vllry** [comment]
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/testing.md this guide might be useful for running some tests locally - it's faster feedback than waiting for Prow.

File: test/e2e/network/network_policy.go (hunk 5 above)
Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: Just updated. Thanks!

**Comment by vllry** [comment]
/test pull-kubernetes-e2e-gce
That one looks like an infra flake.

File: test/e2e/network/network_policy.go [hunk 7]
```diff
@@ -312,6 +314,53 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+			}
+
+			policy, err = f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Create(policy)
+			framework.ExpectNoError(err, "Error creating Network Policy %v: %v", policy.ObjectMeta.Name, err)
+			defer cleanupNetworkPolicy(f, policy)
+
+			ginkgo.By("Creating client-a which should not be able to contact the server.", func() {
+				testCannotConnect(f, f.Namespace, "client-a", service, allowedPort)
+			})
+			ginkgo.By("Creating client-b which should be able to contact the server on port 80.", func() {
```
Comment (vllry) [inline test/e2e/network/network_policy.go]: Magic number snuck in there still ;)

File: test/e2e/network/network_policy.go [hunk 8]
```diff
@@ -312,6 +314,53 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+						}},
+					}},
+				},
+			}
+
+			policy, err = f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Create(policy)
+			framework.ExpectNoError(err, "Error creating Network Policy %v: %v", policy.ObjectMeta.Name, err)
+			defer cleanupNetworkPolicy(f, policy)
+
+			ginkgo.By("Creating client-a which should not be able to contact the server.", func() {
```
Comment (vllry) [inline test/e2e/network/network_policy.go]: This can be removed, as testCannotConnect will add it.

File: test/e2e/network/network_policy.go [hunk 9]
```diff
@@ -353,6 +402,197 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+							Port: &intstr.IntOrString{IntVal: clientAAllowedPort},
+						}},
+					}},
+				},
+			}
+			policy, err := f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Create(policy)
+			framework.ExpectNoError(err, "Error creating Network Policy %v: %v", policy.ObjectMeta.Name, err)
+
+			testCanConnect(f, f.Namespace, "client-a", service, clientAAllowedPort)
+			err = f.WaitForPodNotFound("client-a", framework.PodDeleteTimeout)
```
Comment (vllry) [inline test/e2e/network/network_policy.go]: Would it be faster to use unique names, rather than waiting for the teardown to reuse the name? (I suspect so)

Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: Yes, it would be faster, but it would not cover the corner cases generated when updating a policy rule to use a different `podSelector`. The name given to a pod is also used for labeling it. Thus, I believe is good to ensure that, for example, when the `podSelector` on a `NetworkPolicyIngressRule` is updated the pod that was previous selected is not selected anymore.

Comment (vllry) [reply test/e2e/network/network_policy.go]: I see. A bit unfortunate, but not worth blocking on.

**Comment by vllry** [comment]
/lgtm

@johnbelamaric ready for a look :)

File: test/e2e/network/network_policy.go [hunk 10]
```diff
@@ -353,6 +398,197 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+			if err != nil {
+				framework.ExpectNoError(err, "Expected pod to be not found.")
+			}
+			testCannotConnect(f, f.Namespace, "client-b", service, clientAAllowedPort)
+			err = f.WaitForPodNotFound("client-b", framework.PodDeleteTimeout)
+			if err != nil {
+				framework.ExpectNoError(err, "Expected pod to be not found.")
+			}
```
Comment (oomichi) [inline test/e2e/network/network_policy.go]: The above 3 lines (line 444-446) can be merged into a single `framework.ExpectNoError(err, "Expected pod to be not found.")`

File: test/e2e/network/network_policy.go [hunk 11]
```diff
@@ -353,6 +398,197 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+
+			testCannotConnect(f, f.Namespace, "client-a", service, clientANotAllowedPort)
+			err = f.WaitForPodNotFound("client-a", framework.PodDeleteTimeout)
+			if err != nil {
+				framework.ExpectNoError(err, "Expected pod to be not found.")
+			}
```
Comment (oomichi) [inline test/e2e/network/network_policy.go]: ditto: for line 450-452 into a single call.

File: test/e2e/network/network_policy.go [hunk 12]
```diff
@@ -353,6 +398,197 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+			policy, err = f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Update(policy)
+			framework.ExpectNoError(err, "Error updating Network Policy %v: %v", policy.ObjectMeta.Name, err)
+			defer cleanupNetworkPolicy(f, policy)
+
+			testCannotConnect(f, f.Namespace, "client-b", service, clientBNotAllowedPort)
+			err = f.WaitForPodNotFound("client-b", framework.PodDeleteTimeout)
+			if err != nil {
+				framework.ExpectNoError(err, "Expected pod to be not found.")
+			}
```
Comment (oomichi) [inline test/e2e/network/network_policy.go]: ditto: for line 491-493.

File: test/e2e/network/network_policy.go [hunk 13]
```diff
@@ -353,6 +398,197 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+				},
+			}
+			policy, err := f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Create(policy)
+			framework.ExpectNoError(err, "Error creating Network Policy %v: %v", policy.ObjectMeta.Name, err)
+
+			testCanConnect(f, f.Namespace, "client-a", service, clientAAllowedPort)
+			err = f.WaitForPodNotFound("client-a", framework.PodDeleteTimeout)
+			if err != nil {
+				framework.ExpectNoError(err, "Expected pod to be not found.")
+			}
```
Comment (oomichi) [inline test/e2e/network/network_policy.go]: The above 3 lines (line 439-441) can be merged into a single framework.ExpectNoError(err, "Expected pod to be not found.")

File: test/e2e/network/network_policy.go [hunk 14]
```diff
@@ -574,6 +816,30 @@ func createNetworkClientPod(f *framework.Framework, namespace *v1.Namespace, pod
...
+		Path:  patchPath,
+		Value: patchValue,
+	}}
+	payloadBytes, err := json.Marshal(payload)
+	framework.ExpectNoError(err)
+
+	pod, err := f.ClientSet.
+		CoreV1().
+		Pods(namespace.Name).
+		Patch(podName, types.JSONPatchType, payloadBytes)
```
Comment (oomichi) [inline test/e2e/network/network_policy.go]: nit: Why we need to separate this single API call into multiple lines? Just curious because I never see this pattern on a single API call.

Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: Just followed the pattern I've seen in another file. But, it can avoid having a big line, which is not actually the case in here, so I've updated to a single line.

**Comment by MaysaMacedo** [comment]
Thanks for the review @oomichi  .

**Review by oomichi — COMMENTED** [review]
Thanks for updating.
As https://kubernetes.slack.com/archives/CGXR5GX6Z/p1560200302009800?thread_ts=1560197588.009100&cid=CGXR5GX6Z @MaysaMacedo has confirmed this test works fine on local environment.

/lgtm

**Comment by MaysaMacedo** [comment]
@vllry @oomichi Could you take another look in this PR? I had to do a rebase. Thanks!

**Comment by oomichi** [comment]
Thanks

/lgtm

**Comment by bowei** [comment]
cc: @lzang

File: test/e2e/network/network_policy.go [hunk 15]
```diff
@@ -491,6 +493,49 @@ var _ = SIGDescribe("NetworkPolicy", func() {
...
+			nsBName := f.BaseName + "-b"
+			nsB, err := f.CreateNamespace(nsBName, map[string]string{
+				"ns-name": nsBName,
+			})
+			framework.ExpectNoError(err, "Error creating namespace %v: %v", nsBName, err)
+
+			const allowedPort = 80
+			policy := &networkingv1.NetworkPolicy{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "allow-client-a-via-named-port-ingress-rule",
```
Comment (aanm) [inline test/e2e/network/network_policy.go]: Seems odd the name describes that client-a should be allowed but on line 535 it is testing that client-a should not connect. I guess the policy name is the incorrect part here.

File: test/e2e/network/network_policy.go [hunk 16]
```diff
@@ -753,6 +989,27 @@ func createNetworkClientPod(f *framework.Framework, namespace *v1.Namespace, pod
 	return pod
 }
 
+// Patch client pod with a map value
+func updateNetworkClientPodLabel(f *framework.Framework, namespace *v1.Namespace, podName string, patchOperation string, patchPath string, patchValue map[string]string) *v1.Pod {
+	type patchMapValue struct {
+		Op    string            `json:"op"`
+		Path  string            `json:"path"`
+		Value map[string]string `json:"value"`
```
Comment (aanm) [inline test/e2e/network/network_policy.go]: Value can be omitted for some operations. consider having `json:"value,omitempty"` instead

**Review by aanm — CHANGES_REQUESTED** [review]
2 small nits, very good addition @MaysaMacedo

File: test/e2e/network/network_policy.go [hunk 17]
```diff
@@ -491,6 +493,49 @@ var _ = SIGDescribe("NetworkPolicy", func() {
 			})
 		})
 
+		ginkgo.It("should allow ingress access from namespace on one named port [Feature:NetworkPolicy]", func() {
+			nsBName := f.BaseName + "-b"
+			nsB, err := f.CreateNamespace(nsBName, map[string]string{
+				"ns-name": nsBName,
+			})
+			framework.ExpectNoError(err, "Error creating namespace %v: %v", nsBName, err)
```
Comment (lzang) [inline test/e2e/network/network_policy.go]: Shouldn't we clean up the namespace when the test finished?

Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: Namespaces are being deleted after test is completed.
[kubernetes/test/e2e/framework/framework.go](https://github.com/kubernetes/kubernetes/blob/a404238d1c90ee00827624249b9223e7644e3f99/test/e2e/framework/framework.go#L404)

File: test/e2e/network/network_policy.go (hunk 15 above)
Comment (aanm) [reply test/e2e/network/network_policy.go]: @lzang right, so it should be `allow-client-b-via-named-port-ingress-rule`

**Comment by MaysaMacedo** [comment]
Thanks for the review @aanm @lzang .

**Comment by lzang** [comment]
Thanks Maysa! 
/lgtm

**Review by aanm — APPROVED** [review]

**Comment by cmluciano** [comment]
Need an approver 

/assign @danwinship @caseydavenport
